File: android/app/src/main/kotlin/org/gaindrive/android/ui/browse/ArtistsViewModel.kt
package org.gaindrive.android.ui.browse

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.navigation.toRoute
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import org.gaindrive.android.data.FetchMonitor
import org.gaindrive.android.data.FetchStatus
import org.gaindrive.android.data.LibraryRepository
import org.gaindrive.android.data.ServerFailure
import org.gaindrive.android.data.ServerSelection
import org.gaindrive.android.data.model.BrowseScope
import org.gaindrive.android.data.model.LibraryListing
import org.gaindrive.android.data.model.ServerConfig
import org.gaindrive.android.data.model.ServerId
import org.gaindrive.android.net.runCatchingCancellable
import org.gaindrive.android.net.userMessage
import org.gaindrive.android.ui.Load
import org.gaindrive.android.ui.Route
import javax.inject.Inject

@HiltViewModel
class ArtistsViewModel @Inject constructor(
	private val library: LibraryRepository,
	private val selection: ServerSelection,
	monitor: FetchMonitor,
	savedStateHandle: SavedStateHandle,
) : ViewModel() {

	/**
	 * Whether this instance is the uploads listing rather than the library.
	 * From the route, so the pushed `Route.Uploads` host and the Library tab
	 * share every line of this class — the listing they load is the whole
	 * difference.
	 */
	val uploads: Boolean = savedStateHandle.toRoute<Route.Artists>().uploads

	/**
	 * What is being fetched, for the uploads listing's own row.
	 *
	 * Here as well as in the shell's strip because this is where somebody looks
	 * when they wonder whether a fetch is running — the report that prompted all
	 * of this was "came back to my uploads folder and saw nothing in progress".
	 * The strip answers it from every screen; this answers it at the place the
	 * question is actually asked.
	 *
	 * The listing itself needs nothing: a finished fetch bumps [LibraryRevision],
	 * which ServerSelection.browse already folds into the value collected below.
	 */
	val fetches: StateFlow<FetchStatus> =
		monitor.status.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), FetchStatus())

	/**
	 * A plain held value, not a `stateIn(WhileSubscribed(…))` over the query.
	 *
	 * That was the earlier shape and it re-fetched the whole artist list every
	 * time the screen came back into view: leaving for an album dropped the last
	 * subscriber, the share stopped, and returning restarted the upstream. The
	 * list does not change while the user is two screens deep, so it is simply
	 * kept until something asks for it again.
	 */
	private val _state = MutableStateFlow<Load<LibraryListing>>(Load.Loading)
	val state: StateFlow<Load<LibraryListing>> = _state.asStateFlow()

	/** Servers that did not answer this load, if any did. */
	private val _failures = MutableStateFlow<List<ServerFailure>>(emptyList())
	val failures: StateFlow<List<ServerFailure>> = _failures.asStateFlow()

	/** True only for a user-initiated pull, which drives the pull indicator. */
	private val _isRefreshing = MutableStateFlow(false)
	val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

	/**
	 * Whether the upload icon is worth drawing. Never true on the uploads
	 * listing itself — the icon is how you get there.
	 */
	private val _canUpload = MutableStateFlow(false)
	val canUpload: StateFlow<Boolean> = _canUpload.asStateFlow()

	private var scope: BrowseScope = BrowseScope.AllServers
	private var loadJob: Job? = null

	init {
		viewModelScope.launch {
			// distinctUntilChanged because the registry re-emits whenever
			// anything in DataStore changes, and an unchanged scope is not a
			// reason to re-read the library.
			selection.browse.distinctUntilChanged().collect { selected ->
				scope = selected.scope
				refreshCanUpload()
				// A different scope is a different library, and going offline
				// is the same library from a different source — either way the
				// old list must go rather than linger under a spinner.
				startLoad(clearFirst = true)
			}
		}
	}

	/**
	 * Failure keeps the last answer rather than hiding the icon: it is
	 * navigation, and taking it away because one request timed out would strand
	 * the user out of their own uploads.
	 */
	private fun refreshCanUpload() {
		if (uploads) return
		viewModelScope.launch {
			runCatchingCancellable { library.canUpload(scope) }
				.getOrNull()?.let { _canUpload.value = it }
		}
	}

	/** Retry after a failure: there is nothing worth keeping on screen. */
	fun load() = startLoad(clearFirst = true)

	/** Pull to refresh: keep the list visible while it is re-read. */
	fun refresh() {
		_isRefreshing.value = true
		startLoad(clearFirst = false)
	}

	private fun startLoad(clearFirst: Boolean) {
		loadJob?.cancel()
		loadJob = viewModelScope.launch {
			if (clearFirst) _state.value = Load.Loading
			if (selection.hasNoServers()) {
				_state.value = Load.Failed("No server configured. Add one in Settings.")
				_isRefreshing.value = false
				return@launch
			}

			val loaded = runCatchingCancellable {
				if (uploads) {
					// The personal slice has no categories by construction, so
					// it is an artists-only listing of the same shape.
					library.uploadIndexes(scope).map { LibraryListing(emptyList(), it) }
				} else {
					library.libraryListing(scope)
				}
			}
			loaded.fold(
				onSuccess = { merged ->
					// Every server failing is a failed screen; some of them
					// failing is a note over the ones that worked.
					if (merged.items.isEmpty && merged.isPartial) {
						_state.value = Load.Failed(merged.failures.first().message)
					} else {
						_state.value = Load.Ready(merged.items)
					}
					_failures.value = merged.failures
				},
				onFailure = {
					_state.value = Load.Failed(it.userMessage())
					_failures.value = emptyList()
				},
			)
			_isRefreshing.value = false
		}
	}

	fun dismissFailures() {
		_failures.value = emptyList()
	}

	val servers: StateFlow<List<ServerConfig>> = selection.available
		.stateIn(viewModelScope, SharingStarted.Lazily, emptyList())

	val browseScope: StateFlow<BrowseScope> = selection.scope
		.stateIn(viewModelScope, SharingStarted.Lazily, BrowseScope.AllServers)

	val badgeNames: StateFlow<Map<ServerId, String>> = selection.badgeNames
		.stateIn(viewModelScope, SharingStarted.Lazily, emptyMap())

	fun selectServer(id: ServerId) = viewModelScope.launch { selection.select(id) }

	fun selectAllServers() = viewModelScope.launch { selection.selectAllServers() }

	val offline: StateFlow<Boolean> = selection.offline
		.stateIn(viewModelScope, SharingStarted.Lazily, false)

	fun setOffline(enabled: Boolean) = viewModelScope.launch { selection.setOffline(enabled) }
}

File: ios/GainDrive/UI/Browse/ArtistsViewModel.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

/// The merged library listing - or, with `uploads` set, the account's own
/// uploads - mirroring `ui/browse/ArtistsViewModel.kt`.
///
/// **State is a plain held value**, not something derived from a subscription.
/// Android reversed away from `stateIn(WhileSubscribed)` here because leaving
/// for an album dropped the last subscriber, the share stopped, and returning
/// restarted the upstream - a full re-read of the library every time the screen
/// came back into view. SwiftUI has the identical trap wearing a different hat:
/// `.task` runs again on every re-appearance. `loadedFor` is what closes it.
///
/// **The initialiser stores references and does nothing else.** `RootView.init`
/// re-runs whenever `GainDriveApp.body` re-evaluates - which
/// `preferredColorScheme` guarantees on every theme change - and `@State` keeps
/// the first instance and discards the rest. Any work started in an initialiser
/// would happen once per discarded copy.
@MainActor
@Observable
final class ArtistsViewModel {
	private(set) var state: Load<LibraryListing> = .loading
	private(set) var failures: [ServerFailure] = []
	private(set) var isRefreshing = false
	private(set) var badgeNames: [ServerId: String] = [:]

	/// Whether the upload icon is worth drawing. Never true on the uploads
	/// instance - the icon is how you get there. A failed check leaves the
	/// last answer standing: the icon is navigation, and taking it away
	/// because one request timed out would strand the user out of their own
	/// uploads.
	private(set) var canUpload = false

	/// Whether this instance is the uploads listing rather than the library.
	/// The listing it loads is the whole difference - the pushed uploads
	/// screen and the Library tab share every other line of this class.
	let uploads: Bool

	@ObservationIgnored private let library: LibraryRepository
	@ObservationIgnored private let selection: ServerSelection
	/// What is currently on screen. Leaving for an album and coming back is not
	/// a reason to re-read the library: the list has not changed while the user
	/// was two screens deep.
	@ObservationIgnored private var loadedFor: BrowseScope?
	@ObservationIgnored private var task: Task<Void, Never>?

	init(library: LibraryRepository, selection: ServerSelection, uploads: Bool = false) {
		self.library = library
		self.selection = selection
		self.uploads = uploads
	}

	/// Called from `.task(id:)`, which fires on appearance *and* on a scope
	/// change. The guard is what distinguishes them.
	func appear() {
		let scope = selection.scope
		guard loadedFor != scope else { return }
		start(clearFirst: loadedFor != nil)
	}

	func retry() {
		start(clearFirst: true)
	}

	/// `.refreshable` awaits this, or the pull indicator snaps back before the
	/// request has even left. A refresh keeps the list on screen - blanking it
	/// would hide the very thing the user pulled to update.
	func refresh() async {
		isRefreshing = true
		start(clearFirst: false)
		await task?.value
		isRefreshing = false
	}

	private func start(clearFirst: Bool) {
		let scope = selection.scope
		loadedFor = scope
		if clearFirst { state = .loading }
		badgeNames = selection.badgeNames

		// Fire-and-forget beside the load, never awaited before it: the icon
		// arriving a beat after the list costs nothing, while serialising the
		// two would put an account lookup in front of every listing.
		if !uploads {
			Task { [library] in
				self.canUpload = await library.canUpload(scope: scope)
			}
		}

		task?.cancel()
		// The counterpart of `viewModelScope.launch`: the *view model* owns the
		// task, so navigating away does not cancel the load and strand the
		// screen on `.loading` forever. The view's `.task` closure only kicks
		// this off and returns.
		task = Task { [library, uploads] in
			let merged =
				uploads
				// The personal listing has no categories by construction, so
				// it is an artists-only listing of the same shape.
				? await library.uploadIndexes(scope: scope)
					.map { LibraryListing(categories: [], artists: $0) }
				: await library.libraryListing(scope: scope)
			guard !Task.isCancelled else { return }
			self.state = merged.load
			self.failures = merged.failures
		}
	}

	func dismissFailures() {
		failures = []
	}

	func artistUi(_ artist: Artist) -> ArtistUi {
		ArtistUi(artist: artist, badges: artist.sources.compactMap { badgeNames[$0] })
	}
}
